ibis/backends/pyspark/compiler.py

```diff
@@ -819,16 +820,19 @@ def compile_capitalize(t, expr, scope, timecontext, **kwargs):
 @compiles(ops.Substring)
 def compile_substring(t, expr, scope, timecontext, **kwargs):
     op = expr.op()
-
-    @F.udf('string')
-    def substring(s, start, length):
-        end = start + length
-        return s[start:end]
-
     src_column = t.translate(op.arg, scope, timecontext)
-    start_column = t.translate(op.start, scope, timecontext)
-    length_column = t.translate(op.length, scope, timecontext)
-    return substring(src_column, start_column, length_column)
+    start = t.translate(op.start, scope, timecontext, raw=True) + 1
+    length = t.translate(op.length, scope, timecontext, raw=True)
+
+    if isinstance(start, pyspark.sql.Column) or isinstance(
+        length, pyspark.sql.Column
+    ):
+        raise NotImplementedError(
```

Comment:
is this case tested?

Reply:
This is covered by this:

https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_string.py#L216

When catching NotImplemenntedError, the test will be xfailed.